File: py/496.下一个更大元素-i.py
#
# @lc app=leetcode.cn id=496 lang=python3
#
# [496] 下一个更大元素 I
#

# @lc code=start
# 用栈维护递减数列

class Solution:
    def nextGreaterElement(self, nums1, nums2):
        l = []
        re = {}
        for i in nums2:
            while len(l) > 0:
                if i > l[-1]:
                    tmp_i = l.pop()
                    re[tmp_i] = i
                else:
                    break
            l.append(i)
        r = []
        for i in nums1:
            r.append(re.get(i, -1))
        return r
# @lc code=end

nums1 = [3]
nums2 = [1,3,4,2]
# nums1 = [2,4]
# nums2 = [1,2,3,4]
print(Solution().nextGreaterElement(nums1, nums2))

File: golang/496.下一个更大元素-i.go
package main

import "fmt"

/*
 * @lc app=leetcode.cn id=496 lang=golang
 *
 * [496] 下一个更大元素 I
 */

// @lc code=start
func nextGreaterElement(nums1 []int, nums2 []int) []int {
	re := make(map[int]int, len(nums2))
	r := make([]int, 0)
	l := make([]int, 0)
	for _, i := range nums2 {
		for len(l) != 0 {
			curI := l[len(l)-1]
			if curI < i {
				re[curI] = i
				l = l[:len(l)-1]
			} else {
				break
			}
		}
		l = append(l, i)
	}

	for _, i := range nums1 {
		if j, ok := re[i]; ok {
			r = append(r, j)
		} else {
			r = append(r, -1)
		}
	}
	return r
}

// @lc code=end

func main() {
	nums1 := []int{4, 1, 2}
	nums2 := []int{1, 3, 4, 2}
	fmt.Println(nextGreaterElement(nums1, nums2))

}
